accepted/future-releases/static-extension-methods/feature-specification.md

```diff
@@ -46,214 +46,205 @@ extension MyFancyList<T> on List<T> {
+- A post-increment expression of the form `e1.id++` is generally equivalent to `e1.id += 1` (which is `e1.id = e1.id + 1` except that subexpressions of `e1` are not evaluated more than once), but the value of the expression is the value from evaluating `e1.id` before adding `1`. This too, works similarly when `e1` is an extension application:
+  - `E<...>(e).id++` is equivalent to `E<...>(e).id = E<...>(e).id + 1` except that `e` is only evaluated once, and the value of the increment expression is the value of the subexpression `E<...>(e).id` before the addition. Symmetrically for post-decrement.
+  - `E<...>(e)[e1]++` is equivalent to `E<...>(e)[e2] = E<...>(e)[e2] + 1` except that `e` and `e2` are only evaluated once, and the value of the increment expression is the value of the subexpression `E<...>(e)[e2]` before the addition. Symmetrically for post decrement.
 
-### Generic Parameter Inference
+### Implicit Extension Member Invocation
 
-If both the extension and the method is generic, then inference must infer the extension type parameters first, to figure out whether the extension applies, and only then start inferring method type parameters. As mentioned above, the inference is similar to other cases of chained inference. 
+Extension members can be invoked *implicitly* (without mentioning the extension by name) as if they were members of the `on` type of the extension. This is intended as the primary way to use extensions, with explicit extension member invocation as a fallback for cases where the implicit extension resolution doesn't do what the user want. 
 
-Example:
+An implicit extension member invocation occurs for a simple or composite member invocation with a target expression `e` iff there exists a unique *most specific* extension declaration which is *accessible* and *applicable* to the member invocation (see below).
 
-```dart
-extension SuperList<T> on List<T> {
-  R foldRight<R>(R base, R combine(T element, R accumulator)) {
-    for (int i = this.length - 1; i >= 0; i--) {
-      base = combine(this[i], base);
-    }
-    return base;
-  }
-}
-...
-  List<String> strings = ...;
-  int count(String string, int length) => length + string.length;
-  ...
-  var length = strings.foldRight(0, count);
-```
+If `E` is the single most specific accessible and applicable extension for a member invocation *i* with target expression `e`, which would then be a compile-time error by itself,  then we treat the target expression as if it was the extension application of `E` to `e`, and if `E` is generic, also providing the type arguments inferred for `E` in checking that it was applicable. This makes the member invocation equivalent to an explicit extension member invocation. This happens even if the *name* `E` is not accessible, so this is not a syntactic rewrite.
```

Comment:
Again, this doesn't sound quite right for composite invocations with split origins.

Reply:
What is a "composite invocation with split origins"?

It *should* be correct. If it isn't, then some part of the definitions leading here is insufficient.

Since all extension lookup is by member basename, and the applicability rule uses the basename of the interface to see if any extension is applicable, then we will either find one most specific extension for both read and write, or none at all.

Reply:
> Since all extension lookup is by member basename

This is going to need some clarification and tests.  It sounds like you intend all of the member invocations in this program to be errors:

```dart
class A {
  int get x => 3;
}

extension E on A {
  void set x(int _) {}
  void set y(int _) {}

}

extension E2 on A {
  int get y => 3;
}

void main() {
  var a = new A();
  a.x;
  a.x = 4;
  a.x += 3;
  a.y;
  a.y = 4;
  a.y += 3;
}
```

As of 2pm yesterday, all of these except the last one were accepted by all implementations (the last was rejected by the CFE).

Reading the spec it sounds like, on the other hand, the following should be accepted, since both extensions apply, and one is more specific?

```dart
class B {}
class C extends B {}

extension EB on B {
  int get y => 3;
}

extension EC on C {
  int set y = 3;
}

void test() {
  var b = new C();
  // b.x; should be error?
  b.x = 4;
  // b.x += 3; Should be error?

}
```

Reply:
Just validated that as of this morning, the analyzer reports no errors, and the CFE only reports two errors (marked in comments) on the following program:

```dart
class A {
  int get x => 3;
}

extension E on A {
  void set x(int _) {}
  void set y(int _) {}

}

extension E2 on A {
  int get y => 3;
}

class B {}
class C extends B {}

extension EB on B {
  int get y => 3;
}

extension EC on C {
  void set y(int _) {}
}

void test() {
  var b = new C();
  b.y; // Should be an error?
  b.y = 4; // Should be accepted?
  b.y += 3; // Should be error?  CFE yes

}

void main() {
  var a = new A();
  a.x; 
  a.x = 4; // Should be an error?
  print(a.x += 3); // Should be an error?
  a.y; // Should be an error?
  a.y = 4; // Should be an error?
  print(a.y += 3); // Should be an error? CFE yes
}
```